### DISCUSSION TITLE
Can you encrypt mail data on rest?

### DISCUSSION THREAD

**Discussion by Jarrodsz** [docker-mailserver/docker-mailserver#3186] [Q&A]
Is it possible to encrypt the mail data when its stored on disk?

The reason im asking for this since im looking for as much privacy and security as possibe. I need to handle client data and although mail is not perse the most secure method I do occasional get sensitive financial data attached to mails from clients. I process that and remove it from my mailbox almost instantly but still.

It would be even more awesome if one could only open and retrieve the mails when there is provided an encryption secret only available on the mail client.

A fully encrypted mailbox which cannot be opened without a correct encryption key beside username and password would be what im looking at.

Not sure if this is even possible with the imap protocol anno 2023.  Thank you for all the efforts that went into this repro, been folowwing it dor 4 years already🙏

**Comment by polarathene** [comment]
> Is it possible to encrypt the mail data when its stored on disk?

Dovecot doe support this. DMS presently does not offer easy config options to handle that for you.

I do have a WIP feature that implements this almost ready to go, I just need to find time to complete it and get it merged. If all goes well, that'd happen next month for a future release, so you'd still be waiting a while for official support.

FWIW, you may find it much easier to use a mail service instead of self-hosting DMS. I'm not sure if they encrypt at rest, but MXRoute does offer fairly affordable mail services last I knew. I had a black friday special discount of $15 USD annually that allowed unlimited domains and accounts, the only limit was the mail storage itself IIRC.

If you have sufficient mail traffic, you may no longer qualify for free tier relay services like SendGrid, plus usually you're paying for a server and maintaining that and all the other issues that can entail. Less benefit when you're already using a server for self-hosting other software, but worrying less about the various problems you can encounter with self-hosting mail, you may find a service like MXRoute meets your business needs better (they provide webmail frontends for free too).

---

It's been a while since I last worked on the mail encryption feature, but IIRC, it has the options of a global or per-user encryption key, where per-user needs the user to login to derive the encryption key (admin unable to decrypt).

That provides encrypted mailbox storage, while TLS provides encryption in transit. Anything on the client-side is not encrypted, for that I think they'd have to encrypt the mail itself which is less common.

**Reply by Jarrodsz** [reply]
Sounds awesome hope it will see the light of day sometime. In the meantime have to do without then seems like this was quite a task reading from your response.

**Reply by Jarrodsz** [reply]
At least running own mailserver is a lot better than hosting it on a provider.

My existing provider enabled webmail.provider.com without 2 way auth so basically anyone could try and login using your email there.

As long as server not comprimised nobody could read your mails, no auth could easily seize it.

Encryption would only take this a step further making reading the mails when compromised be impossible if the decryption salts wouls not be on the server but oy on the client.

Wonder how protonmail does it
